src/garage/_primitives.py

```diff
@@ -0,0 +1,152 @@
+"""Policy interface."""
+import abc
+
+
+class Primitive(abc.ABC):
+    """Primitive interface.
+
+    Args:
+        name (str): Name for logs, variable scopes, etc.
+        spec (garage.InOutSpec): Input/output specification this primitive will
+            satisfy.
+
+    """
+
+    def __init__(self, name, spec):
+        self._name = name
+        self._spec = spec
+
+    @abc.abstractmethod
+    def get_parameters(self):
```

Comment:
Moving forward, I think it would actually be better to replace these methods with a protocol, so that outside authors only need to implement e.g. two methods, not four.

Reply:
hmm how would a protocol save them implementing 2 methods?

Reply:
If the author is writing e.g. a `ValueFunction` using pytorch, they need to define `forward` and `compute_loss` (two methods). If we make a protocol that attempts to use `get_parameters` but falls back to `get_state_dict`, then the author doesn't need to know about `get_parameters`.

Reply:
oh hmm -- we would implement that in this class?

Reply:
Yeah, we would have a function (e.g. `garage.get_parameters`) that implements this protocol by attempting to use `get_parameters`, then falling back to `get_state_dict`.

Reply:
Under two cases we might want to get the parameters of a primitives - a) get trainable params for optimizer, b) get all params for snapshot. Case b) should be covered by __setstate__ and __getstate__ so that leaves only case a). Then we need to make sure the usage of `get_parameters` through its name or documents.

Reply:
I disagree that this method should return only trainable parameters. Which parameters are trainable is already better handled by the specific neural network library code. I think there's actually a third case c) for updating a policy with an existing name (e.g. what's usually done in the samplers). In order to handle that use case, this method should return all parameters (not just the trainable ones). Having said that, I still think we should have a protocol instead, so we can e.g. use `torch.nn.Module.get_state_dict` if that's defined and `get_parameters` isn't.

Reply:
I agree. If we run with the presumption that the Model<-->Optimizer API will always be framework-specific (e.g. no comingling of TF models with numpy optimizers, or at least doing so will require extra work/adapters), then there's no need to deal with trainability in this API. We only need be concerned with ALL parameters.

Alternatively, if we go to make a universal optimizer API, we can revisit this one.

I agree that all of these should basically be protocols and that inheriting from ABCs will be optional and even discouraged. The classes here function to define the interface unambiguously.